lib/gui/pages/main/controllers/flash.js

```diff
@@ -16,19 +16,98 @@
 
 'use strict'
 
+const _ = require('lodash')
 const messages = require('../../../../shared/messages')
 const flashState = require('../../../../shared/models/flash-state')
 const driveScanner = require('../../../modules/drive-scanner')
 const progressStatus = require('../../../modules/progress-status')
 const notification = require('../../../os/notification')
 const exceptionReporter = require('../../../modules/exception-reporter')
 const path = require('path')
+const constraints = require('../../../../shared/drive-constraints')
+const units = require('../../../../shared/units')
+const errors = require('../../../../shared/errors')
 
 module.exports = function (
+  $q,
   $state,
   ImageWriterService,
-  FlashErrorModalService
+  FlashErrorModalService,
+  WarningModalService,
+  DriveSelectorService
 ) {
+  /**
+   * @summary Spawn a confirmation warning modal
+   * @function
+   * @public
+   *
+   * @param {String} message - warning message
+   * @returns {Promise} warning modal promise
+   *
+   * @example
+   * confirmationWarningModal('Hello, World!')
+   */
+  const confirmationWarningModal = (message) => {
+    return WarningModalService.display({
+      confirmationLabel: 'Continue',
+      rejectionLabel: 'Change',
+      description: [
+        message,
+        'Are you sure you want to continue?'
+      ].join(' ')
+    })
+  }
+
+  /**
+   * @summary Display warning tailored to the warning of the current drive-image pair
+   * @function
+   * @public
+   *
+   * @param {Object} drive - drive object
+   * @param {Object} image - image object
+   * @returns {Promise}
+   *
+   * @example
+   * displayTailoredWarning(drive, image).then(() => {
+   *   console.log('Continue pressed')
+   * }).catch(() => {
+   *   console.log('Change pressed')
+   * })
+   */
+  const displayTailoredWarning = (drive, image) => {
+    const messageKeys = {
+      largeDriveSize: constraints.isDriveSizeLarge(drive),
+      unrecommendedDriveSize: !constraints.isDriveSizeRecommended(drive, image)
+    }
+    const messageArgs = {
+      largeDriveSize: {
+        humanDriveSize: units.bytesToClosestUnit(drive.size),
+        drive
+      },
+      unrecommendedDriveSize: {
+        image,
+        drive
+      }
+    }
+    const messageKey = _.findKey(messageKeys, _.identity)
+    const message = messages.warning[messageKey](messageArgs[messageKey])
```

Comment:
IMHO this line looks a little funny, with 4 different variables with `message` in the name, and one of the variables being used in two different places to index into two different objects.

I _think_ I see what you're doing with using `const` and lodash all over the place, but I wonder if something like this might be a little bit more readable?
```js
let message = null
if (constraints.isDriveSizeLarge(drive)) {
    message = messages.warning[largeDriveSize]({humanDriveSize: units.bytesToClosestUnit(drive.size), drive})
} else if (!constraints.isDriveSizeRecommended(drive, image)) {
    message = messages.warning[unrecommendedDriveSize]({image, drive})
}
```
?

Or if you really wanted to keep the const-ness I guess you _could_ do:
```js
const message = constraints.isDriveSizeLarge(drive)
    ? messages.warning[largeDriveSize]({humanDriveSize: units.bytesToClosestUnit(drive.size), drive})
    : (
        !constraints.isDriveSizeRecommended(drive, image)
        ? messages.warning[unrecommendedDriveSize]({image, drive})
        : null
    )
```
but that obviously starts heading into the unreadable-code category ;-)

Or perhaps you could do something like:
```js
const messageValues = [
    [constraints.isDriveSizeLarge(drive), messages.warning[largeDriveSize]({humanDriveSize: units.bytesToClosestUnit(drive.size), drive})],
    [!constraints.isDriveSizeRecommended(drive, image), messages.warning[unrecommendedDriveSize]({image, drive})]
}
// and then use `_.find` etc.
```
which is more similar to the approach you're already using, but gets rid of the separate `messageArgs` variable?

\*shrug\* I dunno really, @jviotti can probably offer better advice when it comes to Javascript best-practices ;-)  Feel free to ignore me.

Reply:
I noticed a bug with my current implementation here and now I think the best option is to just go for the if-then-else